File: backend/utils/performance.js
/**
 * Utilitaires pour optimiser les performances du backend
 */

/**
 * Cache simple en mémoire avec TTL
 */
class MemoryCache {
  constructor(defaultTTL = 3600000) { // 1 heure par défaut
    this.cache = new Map();
    this.defaultTTL = defaultTTL;
  }

  set(key, value, ttl = this.defaultTTL) {
    const expiresAt = Date.now() + ttl;
    this.cache.set(key, { value, expiresAt });
    
    // Nettoyer automatiquement après expiration
    setTimeout(() => {
      if (this.cache.has(key)) {
        const item = this.cache.get(key);
        if (item.expiresAt <= Date.now()) {
          this.cache.delete(key);
        }
      }
    }, ttl);
  }

  get(key) {
    const item = this.cache.get(key);
    if (!item) return null;
    
    if (item.expiresAt <= Date.now()) {
      this.cache.delete(key);
      return null;
    }
    
    return item.value;
  }

  delete(key) {
    this.cache.delete(key);
  }

  clear() {
    this.cache.clear();
  }

  size() {
    return this.cache.size;
  }
}

/**
 * Connection pool pour MongoDB
 */
function createConnectionPool(maxConnections = 10) {
  const pool = [];
  let activeConnections = 0;

  return {
    async acquire() {
      if (activeConnections >= maxConnections) {
        await new Promise(resolve => {
          const check = setInterval(() => {
            if (activeConnections < maxConnections) {
              clearInterval(check);
              resolve();
            }
          }, 100);
        });
      }
      activeConnections++;
      return { release: () => activeConnections-- };
    },
    getStats() {
      return { active: activeConnections, max: maxConnections };
    }
  };
}

/**
 * Batch processing pour les opérations en masse
 */
async function batchProcess(items, batchSize, processor) {
  const results = [];
  for (let i = 0; i < items.length; i += batchSize) {
    const batch = items.slice(i, i + batchSize);
    const batchResults = await Promise.all(batch.map(processor));
    results.push(...batchResults);
  }
  return results;
}

/**
 * Debounce pour les opérations coûteuses
 */
function debounce(func, wait) {
  let timeout;
  return function(...args) {
    clearTimeout(timeout);
    timeout = setTimeout(() => func.apply(this, args), wait);
  };
}

/**
 * Mesurer le temps d'exécution
 */
function measureExecutionTime(fn, label) {
  return async function(...args) {
    const start = process.hrtime.bigint();
    try {
      const result = await fn.apply(this, args);
      const end = process.hrtime.bigint();
      const duration = Number(end - start) / 1000000; // Convertir en ms
      
      if (process.env.NODE_ENV === 'development') {
        console.log(`[Performance] ${label}: ${duration.toFixed(2)}ms`);
      }
      
      return result;
    } catch (error) {
      const end = process.hrtime.bigint();
      const duration = Number(end - start) / 1000000;
      console.error(`[Performance] ${label} failed after ${duration.toFixed(2)}ms:`, error);
      throw error;
    }
  };
}

module.exports = {
  MemoryCache,
  createConnectionPool,
  batchProcess,
  debounce,
  measureExecutionTime,
};

File: mobile-app/lib/utils/performance.dart
/// Utilitaires pour optimiser les performances de l'application mobile
import 'dart:async';

/// Cache simple en mémoire avec TTL
class MemoryCache<K, V> {
  final Map<K, _CacheEntry<V>> _cache = {};
  final Duration defaultTTL;

  MemoryCache({this.defaultTTL = const Duration(hours: 1)});

  void set(K key, V value, {Duration? ttl}) {
    final expiresAt = DateTime.now().add(ttl ?? defaultTTL);
    _cache[key] = _CacheEntry(value, expiresAt);
  }

  V? get(K key) {
    final entry = _cache[key];
    if (entry == null) return null;

    if (DateTime.now().isAfter(entry.expiresAt)) {
      _cache.remove(key);
      return null;
    }

    return entry.value;
  }

  void remove(K key) {
    _cache.remove(key);
  }

  void clear() {
    _cache.clear();
  }

  int get length => _cache.length;
  
  Iterable<K> get keys => _cache.keys;
}

class _CacheEntry<V> {
  final V value;
  final DateTime expiresAt;

  _CacheEntry(this.value, this.expiresAt);
}

/// Debounce pour limiter la fréquence d'exécution
class Debouncer {
  final Duration delay;
  Timer? _timer;

  Debouncer({this.delay = const Duration(milliseconds: 300)});

  void call(void Function() callback) {
    _timer?.cancel();
    _timer = Timer(delay, callback);
  }

  void dispose() {
    _timer?.cancel();
  }
}

/// Throttle pour limiter la fréquence d'exécution
class Throttler {
  final Duration limit;
  DateTime? _lastRun;

  Throttler({this.limit = const Duration(milliseconds: 300)});

  bool canRun() {
    final now = DateTime.now();
    if (_lastRun == null || now.difference(_lastRun!) >= limit) {
      _lastRun = now;
      return true;
    }
    return false;
  }
}

/// Mesurer les performances d'une fonction
Future<T> measurePerformance<T>(
  Future<T> Function() function,
  String label,
) async {
  final stopwatch = Stopwatch()..start();
  try {
    final result = await function();
    stopwatch.stop();
    print('[Performance] $label: ${stopwatch.elapsedMilliseconds}ms');
    return result;
  } catch (e) {
    stopwatch.stop();
    print('[Performance] $label failed after ${stopwatch.elapsedMilliseconds}ms: $e');
    rethrow;
  }
}

/// Batch processing pour les opérations en masse
Future<List<T>> batchProcess<T>(
  List<dynamic> items,
  int batchSize,
  Future<T> Function(dynamic) processor,
) async {
  final results = <T>[];
  for (var i = 0; i < items.length; i += batchSize) {
    final batch = items.sublist(
      i,
      i + batchSize > items.length ? items.length : i + batchSize,
    );
    final batchResults = await Future.wait(batch.map(processor));
    results.addAll(batchResults);
  }
  return results;
}




